File: 388. Longest Absolute File Path.cpp
class Solution {
public:
    int lengthLongestPath(string input) {
        int maxi=0,count=0,ln=1;
        bool isFile=false;
        vector<int> level(200);
        level[0]=0;
        for(int i=0,fin=input.size();i<fin;++i){
            //find which level
            while(input[i]=='\t'){
                ++ln;++i;
            }
            //read file name
            while(input[i]!='\n'&&i<fin){
                if(input[i]=='.')isFile=true;
                ++count;++i;
            }
            //calculate
            if(isFile){
                maxi=max(maxi,level[ln-1]+count);
            }
            else{
                level[ln]=level[ln-1]+count+1;// 1 means '/'
            }
            //reset
            count=0;ln=1;isFile=false;
        }
        return maxi;
    }
};

File: python/388. Longest Absolute File Path.py
class Solution(object):
    def lengthLongestPath(self, input):
        """
        :type input: str
        :rtype: int
        """
        st = [(-1, 0)] # st保存了目录深度和对应的字符串长度
        max_len = 0
        for p in input.split('\n'):
            depth = p.count('\t')
            p = p.replace('\t', '')
            while st and st[-1][0] >= depth:
                st.pop()
            
            if '.' in p:
                max_len = max(max_len, len(p) + st[-1][1])
            else:
                st.append((depth, len(p) + st[-1][1] + 1)) # +1 表示多一个/分隔符
        
        return max_len